sdk/highlight-ruby/highlight/lib/highlight.rb

```diff
@@ -0,0 +1,159 @@
+                HIGHLIGHT_SESSION_ATTRIBUTE => session_id,
+                HIGHLIGHT_TRACE_ATTRIBUTE => request_id,
+            }.compact) do |span|
+                begin
+                    yield
+                rescue => e
+                    record_exception(e)
+                    raise
+                end
+            end
+        end
+
+        def record_exception(e)
+            span = OpenTelemetry::Trace.current_span
+            if !span
+                raise "H.record_exception called without a span context"
+            end
+            span.record_exception(e)
+        end
+
+        def record_log(session_id, request_id, level, message)
+            caller_info = caller[0].split(":", 3)
+            function = caller_info[2]
+            if function
+                # format: "in `<function_name>""
+                function.delete_prefix!("in `")
+                function.delete_suffix!("\"")
+            end
+            @tracer.in_span("highlight-ctx", attributes: { 
+                HIGHLIGHT_PROJECT_ATTRIBUTE => @project_id, 
+                HIGHLIGHT_SESSION_ATTRIBUTE => session_id,
+                HIGHLIGHT_TRACE_ATTRIBUTE => request_id,
+            }.compact) do |span|
+                if level == Logger::ERROR || level == Logger::FATAL
+                    span.status = OpenTelemetry::Trace::Status.error(message)
+                end
+                span.add_event(LOG_EVENT, attributes: {
+                    LOG_SEVERITY_ATTRIBUTE => H.log_level_string(level),
+                    LOG_MESSAGE_ATTRIBUTE => message,
+                    CODE_FILEPATH => caller_info[0],
+                    CODE_LINENO => caller_info[1],
+                    CODE_FUNCTION => function,
+                })
+            end
+        end
+
+        HighlightHeaders = Struct.new("HighlightHeaders", :secureSessionId, :requestId)
+        def self.parse_headers(headers)
+            if headers && headers[HIGHLIGHT_REQUEST_HEADER]
+                secureSessionId, requestId = headers[HIGHLIGHT_REQUEST_HEADER].split("/")
+                return HighlightHeaders.new(secureSessionId, requestId)
+            end
+            return HighlightHeaders.new(nil, nil)
+        end
+
+        private
+        
+        def self.log_level_string(level)
+            case level
+            when Logger::UNKNOWN
+                "UNKNOWN"
+            when Logger::FATAL
+                "FATAL"
+            when Logger::ERROR
+                "ERROR"
+            when Logger::WARN
+                "WARN"
+            when Logger::INFO
+                "INFO"
+            when Logger::DEBUG
+                "DEBUG"
+            else
+                "UNKNOWN"
+            end
+        end
+    end
+
+    class Logger < ::Logger
+        def add(severity, message = nil, progname = nil)
+            # https://github.com/ruby/logger/blob/master/lib/logger.rb
+            severity ||= UNKNOWN
+            if @logdev.nil? or severity < level
```

Comment:
I'm also not sure why we are returning true when `severity < level`....what's the intention of returning early?

Reply:
Thanks for sharing! I understand now what is going on. 
It does seem a bit brittle to recreate the ruby core library code. If they change the code, we have to maintain parity with that. 

From reading the code, I thought that creating our own [`LogDevice`](https://ruby-doc.org/stdlib-2.5.3/libdoc/logger/rdoc/Logger/LogDevice.html) would have been the trick since that's what the ruby code ultimately calls:

https://github.com/ruby/logger/blob/master/lib/logger.rb#L668-L670

Yet we don't get the breakdown of what the severity was.

Reply:
One question though is that is that it seems like we don't call `@logdev.write(` in this overridden code. Given that this is expected to be instantiated with STDOUT: https://github.com/highlight/highlight/blob/350e2b5a10ca063013650e1bbc943fcedf1f3397/sdk/highlight-ruby/e2e/rails/blog/config/initializers/highlight.rb#L4

Does that mean we don't print anything out? Shouldn't we still print this out to stdout?

Reply:
I do call `super` though, and that's defaulting to the Ruby Logger to write to STDOUT